app/controllers/feedbacks_controller.rb

```diff
@@ -0,0 +1,83 @@
+class FeedbacksController < ApplicationController
+  before_action :set_feedback, only: [:show, :edit, :update, :destroy]
+  before_action :set_feedbackable, only: [:new, :create]
+  before_action :set_volunteer, only: [:new, :create]
+
+  def index
+    authorize Feedback
+    @feedbacks = policy_scope(Feedback)
+    if params[:group_offer_id]
+      @feedbacks = @feedbacks.where(feedbackable_id: params[:group_offer_id])
+      @volunteer = current_user.volunteer || @feedbacks.first.volunteer
+    elsif params[:assignment_id]
+      @feedbacks = @feedbacks.where(feedbackable_id: params[:assignment_id])
+      @volunteer = Assignment.find(params[:assignment_id]).volunteer
+    elsif params[:volunteer_id]
+      @feedbacks = @feedbacks.where(volunteer_id: params[:volunteer_id])
+      @volunteer = Volunteer.find(params[:volunteer_id])
+    end
+  end
+
+  def show; end
+
+  def new
+    @feedback = Feedback.new(feedbackable: @feedbackable, volunteer: @volunteer,
+      author: current_user)
+    authorize @feedback
+  end
+
+  def edit; end
+
+  def create
+    @feedback = Feedback.new(feedback_params.merge(author_id: current_user.id))
+    authorize @feedback
+    if @feedback.save
+      redirect_to @feedback.volunteer, make_notice
+    else
+      render :new
+    end
+  end
+
+  def update
+    if @feedback.update(feedback_params)
+      redirect_to @feedback.volunteer, make_notice
+    else
+      render :edit
+    end
+  end
+
+  def destroy
+    @feedback.destroy
+    redirect_back(fallback_location: url_for(@feedbackable))
+  end
+
+  private
+
+  def set_feedbackable
+    return @feedbackable = Assignment.find(params[:assignment_id]) if params[:assignment_id]
+    return @feedbackable = GroupOffer.find(params[:group_offer_id]) if params[:group_offer_id]
+    if params[:feedbackable_type] == 'GroupOffer'
+      @feedbackable = GroupOffer.find(params[:feedbackable_id])
+    end
```

Comment:
This method still seems a bit convoluted but I don't understand the feature well enough to say how it could be simplified :)

Reply:
I'm not yet sure how. The fact that superadmins can also add feedback complicates the whole thing.
Otherwise set_feedbackable and set_volunteer could be half as big.